Add CMake option to enable or disable Flang support

dmartinovic-htec opened the pull request:
## Motivation

Closes #1588

## Technical Details

Introduced a new configuration option in the root CMakeLists.txt to control the enablement of Flang (Fortran frontend) support.

dmartinovic-htec committed:
Added option to enable/disable flang

ronlieb commented on CMakeLists.txt:
```diff
@@ -163,6 +163,7 @@ option(THEROCK_ENABLE_ML_LIBS "Enable building of ML libraries" "${THEROCK_ENABL
 option(THEROCK_ENABLE_PROFILER "Enable building the profiler libraries" "${THEROCK_ENABLE_ALL}")
 option(THEROCK_ENABLE_HOST_MATH "Build all bundled host math libraries by default" OFF)
 option(THEROCK_RESET_FEATURES "One-shot flag which forces all feature flags to their default state for this configuration run" OFF)
+option(THEROCK_ENABLE_FLANG "Enable building Flang (Fortran frontend) as part of LLVM" OFF)
```
i would prefer this defaults to ON and that folks can disable if they wish to do so.  The AMD compiler team is rather invested in developing the Fortran compiler, and we currently have this defaulting to ON.

dmartinovic-htec replied:
Done.

ronlieb reviewed (changes requested):
Nice to have the option to disable/enable Flang build. .Currently flang build is on by default, this PR changes default to OFF. Please change the default back to ON,

marbre reviewed (changes requested):
Options are not propagated to pre-hooks as assumed here. Thus `THEROCK_ENABLE_FLANG` is never set in `pre_hook_amd-llvm.cmake` and flang will always be disabled without the ability to control it properly. This would need a change in `therock_subproject.cmake` to work as desired. Furthermore, if adding a new option, please add it to the README.

dmartinovic-htec committed:
Make flang enable flag globally available

This adds:
- THEROCK_ENABLE_FLANG_SUPPORT
to the global properties injected into the init file of each subproject.

Specifically, THEROCK_ENABLE_FLANG_SUPPORT is required in the AMD LLVM pre_hook

dmartinovic-htec committed:
Remove redundant option definition

benrichard-amd commented:
Would be great if we could get this merged.

umarinkovic committed:
Coupled more variables with THEROCK_ENABLE_FLANG
due to branch update.

umarinkovic committed:
Clean-up project enablement logic

lucbruni-amd commented:
What other than the current merge conflicts and relevant failing CI jobs (if any) is blocking this from being reviewed/merged?

Muhamed-Husic committed:
Fix unintended lines introduced during rebase (deleted in #2268)

stellaraccident commented:
Fyi - I'm close to having this land able: https://github.com/ROCm/TheRock/pull/3928

(Which let's us consistently enable and disable fortran throughout rocm)

Muhamed-Husic committed:
conditionalize Flang runtimes and disable Fortran examples without Flang

Muhamed-Husic committed:
remove hardcoded FORTRT flag to allow dynamic config (cache cleanup)

Muhamed-Husic commented:
Thanks for the update. I was recently tasked with cleaning up this PR and pushed some updates over the last few days to resolve merge conflicts and get build passing. After looking at your #3928  its clear that your approach supersedes what's here. I’ll leave it to you to decide how to proceed with this PR.
